File: tests/src/joda-time/joda-time/2.9.9/src/test/java/joda_time/joda_time/DateMidnightInnerPropertyTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package joda_time.joda_time;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import org.joda.time.DateMidnight;
import org.joda.time.DateTimeZone;
import org.junit.jupiter.api.Test;

public class DateMidnightInnerPropertyTest {

    @SuppressWarnings("deprecation")
    @Test
    void serializesAndDeserializesDayOfMonthProperty() throws Exception {
        DateMidnight.Property original = new DateMidnight(2020, 2, 29, DateTimeZone.UTC).dayOfMonth();

        DateMidnight.Property restored = deserialize(serialize(original));

        assertThat(restored.get()).isEqualTo(original.get());
        assertThat(restored.getFieldType()).isEqualTo(original.getFieldType());
        assertThat(restored.getDateMidnight()).isEqualTo(original.getDateMidnight());
    }

    private static byte[] serialize(Serializable value) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        try (ObjectOutputStream objectOutput = new ObjectOutputStream(output)) {
            objectOutput.writeObject(value);
        }
        return output.toByteArray();
    }

    private static DateMidnight.Property deserialize(byte[] serialized) throws IOException, ClassNotFoundException {
        try (ObjectInputStream objectInput = new ObjectInputStream(new ByteArrayInputStream(serialized))) {
            return (DateMidnight.Property) objectInput.readObject();
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/DateMidnightInnerPropertyTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

import org.joda.time.DateMidnight
import org.joda.time.DateTimeFieldType
import org.joda.time.DateTimeZone
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class DateMidnightInnerPropertyTest {
  @Test
  def serializesAndDeserializesDateMidnightProperty(): Unit = {
    val originalDate: DateMidnight = new DateMidnight(2024, 5, 14, DateTimeZone.UTC)
    val originalProperty: DateMidnight.Property = originalDate.dayOfMonth()

    val serialized: Array[Byte] = serialize(originalProperty)
    val restoredProperty: DateMidnight.Property = deserializeProperty(serialized)

    assertEquals(originalDate, restoredProperty.getDateMidnight)
    assertEquals(DateTimeFieldType.dayOfMonth(), restoredProperty.getField.getType)
    assertEquals(14, restoredProperty.get)
  }

  private def serialize(property: DateMidnight.Property): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val output: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      output.writeObject(property)
    } finally {
      output.close()
    }
    bytes.toByteArray
  }

  private def deserializeProperty(serialized: Array[Byte]): DateMidnight.Property = {
    val input: ObjectInputStream = new ObjectInputStream(new ByteArrayInputStream(serialized))
    try {
      input.readObject().asInstanceOf[DateMidnight.Property]
    } finally {
      input.close()
    }
  }
}
